### PR TITLE
Initial TMOP adaptivity interface [tmop-adaptivity]

### PR DESCRIPTION
**Pull request by vladotomov**
The miniapps have examples of analytic and discrete adaptivity.
The interface is still in tmop.hpp/cpp.
Specific implementations were moved from the miniapps to tmop_tools.hpp/cpp.

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/799 | @vladotomov | @v-dobrev | @tzanio + @kmittal2 + @avary-kolasinski | 12/04/19 | 12/06/19 | ⌛due 12/16/20 |

### CODE REVIEW

File: miniapps/meshing/square01.mesh [hunk 1]
```diff
@@ -0,0 +1,114 @@
+MFEM mesh v1.0
```
Comment (tzanio) [inline miniapps/meshing/square01.mesh]: Can't we use `../../data/inline-quad.mesh` instead of this mesh?

Comment (vladotomov) [reply miniapps/meshing/square01.mesh]: This has boundary attributes for tangential motion.

File: fem/tmop_tools.cpp [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+   K.BilinearForm::operator=(0.0);
+   K.Assemble();
+   Vector rhs(K.Size());
+   K.Mult(ind, rhs);
+   M.BilinearForm::operator=(0.0);
+   M.Assemble();
+
+   di_dt = 0.0;
+   CGSolver lin_solver;
+   //GMRESSolver lin_solver;
```
Comment (tzanio) [inline fem/tmop_tools.cpp]: Do we want to keep this and the other `GMRESSolver` lines commented out?

**Comment by tzanio** [comment]
Any cool pictures from the miniapps for https://mfem.org/gallery ?

**Comment by tzanio** [comment]
I think this is ready for `next` but lets wait a bit for the backlog to clear.

**Comment by tzanio** [comment]
@v-dobrev: this is ready for testing and has been merged in `next`

**Comment by tzanio** [comment]
There is a conflict between this PR and #895. See https://github.com/mfem/mfem/pull/895#issuecomment-562894469.

**Comment by tzanio** [comment]
Re-merged in `next` for testing

**Pull request merged**